### DISCUSSION TITLE
Something about 'Storage Access Framework' changes ...

### DISCUSSION THREAD

**Discussion by wmtj** [airsdk/Adobe-Runtime-Support#2533] [General]
My app has an 'export file to local storage / import file from local storage' feature. (app.xml :  uses-sdk  android:targetSdkVersion="29" )

It works fine on AIRSDK 50.1 and before. On AIRSDK 50.2, it does't work anymore, no file permission request window popup, only 'access denied'.

Does this mean that new 'Storage Access Framework' has completely turned off local file access? Or do I need to make some changes in my code? 

@ajwfrost

**Comment by ajwfrost** [comment]
I guess this depends what you mean by "local storage" .. but to export a file, using Storage Access Framework you should use the `File.browseForSave()` option which inherently grants permissions.

If you want to revert to the earlier behaviour, you can add this into your application descriptor, within the `<android>` section:
```
    <storageAccessFrameworkFromAPI>999</storageAccessFrameworkFromAPI>
```
i.e. it won't switch to using SAF until you're on android version 999...

thanks

**Reply by wmtj** [reply]
Thanks!

**Comment by balaedit** [comment]
AIR SDK : 50.2.2.1

```
import flash.filesystem.*;
import flash.events.Event;
var vFile:File = new File();
vFile.addEventListener(Event.SELECT, dirSelect);
vFile.browseForDirectory("Select the Directory");
function dirSelect(e) {
	mytxt.text = vFile.nativePath;
	saveFile();
}
function saveFile() {
	var usrDeviceIDStream = new FileStream();
	var usrDeviceIDFile: File = vFile.resolvePath("id.txt");
	usrDeviceIDStream.open(usrDeviceIDFile, FileMode.WRITE);
	usrDeviceIDStream.writeUTFBytes("File from as3");
	usrDeviceIDStream.close();
}
```
Browse the directory opened the Android UI where I had tap the allow button though getting access denied while saving file.

Error #3001: File or directory access denied.
	at flash.filesystem::FileStream/open()
	at file_fla::MainTimeline/saveFile()[file_fla.MainTimeline::frame1:40]
	at file_fla::MainTimeline/dirSelect()[file_fla.MainTimeline::frame1:13]

![debug](https://user-images.githubusercontent.com/37571867/226531005-4cb29d0b-5f99-41e2-a4ce-09bff2e6bf34.png)

If possible, while debugging in file object we could add that the current path object is permission given or not. 
For example if _var xyz:File = File.applicationDirectory.resolvePath("temp");_

xyz.isPermission method/parameter could be add in the object will be easy while debugging.

It would easy to share the sample code. Thanks.

**Reply by ajwfrost** [reply]
Hi @balaedit -> interesting case, as you can see from the `nativePath` in your screenshot there, something has gone wrong within the `resolvePath` function. From our trace:
```
Selected a folder -> content://com.android.externalstorage.documents/tree/primary%3ADocuments
Trying to create a file: /content:/com.android.externalstorage.documents/tree/primary%3ADocuments/id.txt
```
So firstly, the `resolvePath` isn't working properly.. but even when we correct that `/content:/` to `content://`, we get:
```
Now trying to create a file: content://com.android.externalstorage.documents/tree/primary%3ADocuments/id.txt
Error: Error #3006: Not a file.
```

Will have to look into this one more :-(

**Reply by balaedit** [reply]
As you said that 50.2.2.2 will release by Wednesday, may I expect in it or to wait for it.

**Reply by ajwfrost** [reply]
No the 50.2.2.2 is going to be just an ADT update for IPA code signing and app store deployments.. this will be in 50.2.2.3, I guess, but we're planning to push out a runtime update in the next week or so...

**Comment by balaedit** [comment]
AIR SDK : 50.2.3.1

When saving the file :
```
var vFile: File = new File();
vFile.addEventListener(Event.SELECT, dirSelect);
vFile.browseForDirectory("Select the Directory");
var usrDeviceIDStream = new FileStream();
var usrDeviceIDFile: File = vFile.resolvePath("id.txt");
usrDeviceIDStream.open(usrDeviceIDFile, FileMode.WRITE);  
usrDeviceIDStream.writeUTFBytes("File from as3");
usrDeviceIDStream.close();
```

In above while intent permission given to vFile object we get `file:///content://com.android.externalstorage.documents/tree/primary%253ATestin`

I am saving the above url vFile object in File.applicationStorageDirectory.resolvePath("fileLoc.txt");
When retrieve the file setting it :
```
var sFile:File = File.applicationStorageDirectory.resolvePath("fileLoc.txt");
var sFileStream:FileStream = new FileStream();
if (sFile.exists) {
	sFileStream.open(sFile, FileMode.READ);
	var str = sFileStream.readUTFBytes(sFileStream.bytesAvailable);
	vFile.url = str;
}
```

here I am getting vFile url : file:///content:/com.android.externalstorage.documents/tree/primary%253ATestin

1st while intent vFile :    `file:///content://com.android.externalstorage.documents/tree/primary%253ATestin`
2nd while setting vFile : `file:///content:/com.android.externalstorage.documents/tree/primary%253ATestin`

there is one forward slash missed while setting. When will resolve. Is there any other option to resolve this issue.

**Reply by balaedit** [reply]
Any update??????

**Reply by ajwfrost** [reply]
I'm not actually sure what you're doing there? It sounds like you're getting a folder name, which has a URL that has `file:///content://com.android.....` and you're saving this into a file. Then you're reading back the same file but now the contents have changed to `file:///content:/com.android.....`?

Or is it that the contents hadn't changed, but if you set `vFile.url` to this value, and then read it back, you get a different result?

We can look into that, but the better option might be to set the `vFile.nativePath` option instead. URLs have specific formats and this Android `file:///content://...` thing doesn't conform to that. It's very much a native (android) mechanism.. I would definitely suggest not using the `url` property of a File object on Android, just stick with `nativePath` the whole time..

thanks

**Reply by balaedit** [reply]
Dear @ajwfrost , 
I will brief the situation of the app requirement.
Below is the code for the final requirement functioning call.
```
function saveFile(fFile) {
    var usrDeviceIDStream = new FileStream();
    var usrDeviceIDFile: File = fFile.resolvePath("id.txt");
    usrDeviceIDStream.open(usrDeviceIDFile, FileMode.WRITE);
    usrDeviceIDStream.writeUTFBytes("File from as3");
    usrDeviceIDStream.close();
}
```

App required the internal path of the mobile as direct File.documentsDirectory.resolvePath("loc") will not work so I used : 

```
var vFile:File = new File();
var data;
function fileAccessCheck() {
      vFile.addEventListener(Event.SELECT, dirSelect);
      vFile.browseForDirectory("Select the Directory");
      data = vFile.url;
}

```

So I saved the vFile.url into the File.applicationStorageDirectory.resolvePath("fileLoc.txt");

```
var sFile:File = File.applicationStorageDirectory.resolvePath("fileLoc.txt");
var sFileStream:FileStream = new FileStream();
sFileStream.open(sFile, FileMode.WRITE);
sFileStream.writeUTFBytes(data);   // data has content of vFile.url
sFileStream.close();
saveFile(vFile);
```

When I recheck the to saved file : 

```
function fileCheck() {
        var sFile:File = File.applicationStorageDirectory.resolvePath("fileLoc.txt");
        var sFileStream:FileStream = new FileStream();
        if (sFile.exists) {
	        sFileStream.open(sFile, FileMode.READ);
	        var str = sFileStream.readUTFBytes(sFileStream.bytesAvailable);
	        pFile.nativePath = str;
	        saveFile(pFile);
        } else {
	        fileAccessCheck();
        }
} 
```

Even if I change the url to the native path we get error : 
```
ArgumentError: Error #2004: One of the parameters is invalid.
	at Error$/throwError()
	at flash.filesystem::File/set nativePath()
	at Function/file1_fla:MainTimeline/but2Fn/file1_fla:fileCheck()[file1_fla.MainTimeline::frame1:20]
	at file1_fla::MainTimeline/but2Fn()[file1_fla.MainTimeline::frame1:11]
```

I also sharing the final file please check. While debugging in animate CC, there is one button in center and then first click we get the intent from browse for directory. Then is saved the location of the browsed directory's url to the app storage directory. When again click to the same button then it get break the script.

[file.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/12197270/file.zip)

**Reply by balaedit** [reply]
@ajwfrost Have you got my issue. At least File.documentsDirectory.resolvePath("documents").resolvePath("temp"); should work but failed creating file in temp folder event after request permission.

**Reply by balaedit** [reply]
@ajwfrost  anyupdate?

**Reply by balaedit** [reply]
https://github.com/airsdk/Adobe-Runtime-Support/discussions/2533#discussioncomment-5378583 

@ajwfrost also check this...

**Comment by balaedit** [comment]
https://github.com/airsdk/Adobe-Runtime-Support/discussions/2533#discussioncomment-5378729

@ajwfrost When I it will be available?

**Reply by ajwfrost** [reply]
Apologies for the delay here .. I'll ask someone to check with the code/clarifications you had provided last week, there are a few smaller Android fixes that we may try to squeeze in to a further update/patch release in the near future.

thanks

**Reply by balaedit** [reply]
Please ... do needful as soon as possible... thanks

**Reply by balaedit** [reply]
Thanks @ajwfrost, with release of AIR SDK AIR 50.2.3.4, the testing file is working properly. Appreciated and thanks again for resolving the issue.

**Reply by ajwfrost** [reply]
Thanks for confirming!
